packages/dds/merge-tree/package.json

```diff
@@ -105,11 +105,8 @@
   },
   "typeValidation": {
     "version": "2.0.0-internal.2.2.0",
-    "baselineRange": "2.0.0-internal.2.0.0",
-    "broken": {
-      "ClassDeclaration_Client": {
-        "forwardCompat": false
-      }
-    }
+    "baselineRange": ">=2.0.0-internal.2.1.0 <2.0.0-internal.2.2.0",
+    "baselineVersion": "2.0.0-internal.2.1.0",
+    "broken": {}
```

Comment:
to confirm my understanding of this change:

- #12660 introduced a new API to `Client`, and was merged into `main` after the `2.0.0-internal.2.0.0` release branch was cut but before `2.0.0-internal.2.1.0`. Thus, customers would see forward compatibility issues if they tried to assign a `2.0.0-internal.2.0.0` Client to a `2.0.0-internal.2.1.0` Client.
- This PR adjusts type validation such that `main` runs type validation against the most recent release branch (call it N), N runs type validation against N-1, N-1 against N-2, etc. (the situation is more nuanced for other release groups / next / etc which is why there's a "fluidBuild" config setting the type of version validation to do, but this is the relevant part for this change)
- Due to the above strategy, this change now causes this "typeValidation" section to track which breaking changes have been introduced between 2.0.0-internal.2.1.0 and 2.0.0-internal.2.2.0, of which there are none.

Reply:
this new strategy also gives us better protection for cases like:

version 2.1.0: new API "foo" introduced
version 2.2.0: API "foo" removed

since if "main" was running type validation against 2.0.0, it wouldn't have seen "foo" at all and so not seen removal of "foo" as a breaking change. Right?

Reply:
> This PR adjusts type validation such that main runs type validation against the most recent release branch (call it N), N runs type validation against N-1, N-1 against N-2, etc. (the situation is more nuanced for other release groups / next / etc which is why there's a "fluidBuild" config setting the type of version validation to do, but this is the relevant part for this change)

This isn't how I think of it (I tend to think in semver terms), but yes, I think in practice this is how it plays out. main is supposed to baseline the latest patch of the previous minor, which as you noted, should be the latest release branch. And you're correct about the reason there's a per-branch/per-release group config.

> Due to the above strategy, this change now causes this "typeValidation" section to track which breaking changes have been introduced between 2.0.0-internal.2.1.0 and 2.0.0-internal.2.2.0, of which there are none.

Yes! I'm happy this was clear enough that you made sense of it. :) The typeValidation section is essentially a record of breaking type changes between the baseline version and the repo, which for main should mean it's empty. The baselines will be set according to this table (taken from the docs):

| Branch release designation | Baseline version | Example: version 2.3.4 |
| -------------------------- | ---------------- | ---------------------- |
| `patch`                    | `previousPatch`  | **2.3.3**              |
| `minor`                    | `^previousMinor` | **~2.2.0**             |
| `major`                    | `^previousMajor` | **^1.0.0**             |

Once this change has made it to next, there will be additional broken type overrides that will need to be added, because the baseline will go to an earlier version.

Reply:
> since if "main" was running type validation against 2.0.0, it wouldn't have seen "foo" at all and so not seen removal of "foo" as a breaking change. Right?

Yeah, that's correct. And we get coverage that we don't introduce something in a patch via the release branches, which will baseline to the last released patch.

And the next baseline should make sure we have to explicitly call out breaks between the latest minor release and the next major. We should already be doing that, but this will help enforce.

Reply:
> The typeValidation section is essentially a record of breaking type changes between the baseline version and the repo, *which for main should mean it's empty*

Should it really be completely empty for `main`? Back-compat breaking changes should not appear here, that part makes sense, but forward-compat changes are still possible from one minor version to another, right? Adding a new property to a class, for example.